### PR TITLE
Routing: Fix try_count verification

### PR DESCRIPTION
**Pull request by bigmontz**
The drivers differs a bit about when it should or not call the transaction function.
The java driver only call the transaction function if the connection is fully setup, meanwhile the JS driver delay the connection resolution until the last moment.

Thus the try_count is not a good way to indirect check if the driver try to call all the servers needed. Ideally, It should have a way to check if the full script was called, but since it's not possible at this point, the adopt alternative was checking for the occurrence of some events which could indicates the script was run on the stub server execution log.

### CODE REVIEW

File: tests/stub/routing.py [hunk 1]
```diff
@@ -1128,8 +1128,14 @@ def work(tx):
...
         self._writeServer2.done()
+
+        self.assertGreaterEqual(
+            self._routingServer1.count_responses("SUCCESS"), 2)
+        self.assertGreaterEqual(
+            self._writeServer1.count_responses("<EXIT>"), 1)
+        self.assertGreaterEqual(
+            self._writeServer2.count_responses("SUCCESS"), 4)
         self.assertEqual([[], []], sequences)
-        self.assertEqual(2, num_retries)
```
Comment (injectives) [inline tests/stub/routing.py:1132]: Should we also remove the `num_retries` declaration?

**Comment by robsdedude** [comment]
You can check if the whole stub script was played. `server.done()` will raise an exception if there is the script is unfinished.
Note: if `<EXIT>` is invoked at any point in the script, this counts as finishing the script. Especially noteworthy, it `!AUTO: GOODBYE` is present, the client can reach the end of the script at any time by sending a `GOODBYE` message.
Furthermore, for `!ALLOW RESTART` and `!ALLOW CONCURRENT`, `server.done()` will check that all open connections have reached the end of the script. This means, if the client never connects to the server, this also counts as having finished the script.

File: tests/stub/routing.py [hunk 2]
```diff
@@ -1128,8 +1128,14 @@ def work(tx):
         self._routingServer1.done()
         self._writeServer1.done()
         self._writeServer2.done()
+
+        self.assertGreaterEqual(
+            self._routingServer1.count_responses("SUCCESS"), 2)
+        self.assertGreaterEqual(
+            self._writeServer1.count_responses("<EXIT>"), 1)
+        self.assertGreaterEqual(
+            self._writeServer2.count_responses("SUCCESS"), 4)
```
Comment (robsdedude) [inline tests/stub/routing.py]:
```
        self.assertGreaterEqual(
            self._routingServer1.count_responses("SUCCESS"), 2
        )
```
...

File: tests/stub/routing.py [hunk 3]
```diff
@@ -1564,8 +1570,13 @@ def work(tx):
         self._readServer1.done()
         self._readServer2.done()
         self._readServer3.done()
+
+        self.assertGreaterEqual(self._routingServer1.count_responses("SUCCESS"), 2)
+        self.assertGreaterEqual(self._routingServer2.count_responses("SUCCESS"), 2)
+        self.assertGreaterEqual(self._readServer1.count_responses("<EXIT>"), 1)
+        self.assertGreaterEqual(self._readServer2.count_responses("RECORD"), 1)
+        self.assertGreaterEqual(self._readServer3.count_responses("<EXIT>"), 1)
```
Comment (robsdedude) [inline tests/stub/routing.py]: Try to keep the lines shorter or equal to 80 characters long.

File: tests/stub/routing.py [hunk 4]
```diff
@@ -1599,8 +1610,12 @@ def work(tx):
         self._writeServer1.done()
         self._writeServer2.done()
         self._writeServer3.done()
+        self.assertGreaterEqual(self._routingServer1.count_responses("SUCCESS"), 2)
+        self.assertGreaterEqual(self._routingServer2.count_responses("SUCCESS"), 2)
+        self.assertGreaterEqual(self._writeServer1.count_responses("<EXIT>"), 1)
+        self.assertGreaterEqual(self._writeServer2.count_responses("SUCCESS"), 4)
+        self.assertGreaterEqual(self._writeServer3.count_responses("<EXIT>"), 1)
```
Comment (robsdedude) [inline tests/stub/routing.py]: 80 characters

**Review by robsdedude — CHANGES_REQUESTED** [review]

**Comment by bigmontz** [comment]
> You can check if the whole stub script was played. `server.done()` will raise an exception if there is the script is unfinished.
> Note: if `<EXIT>` is invoked at any point in the script, this counts as finishing the script. Especially noteworthy, it `!AUTO: GOODBYE` is present, the client can reach the end of the script at any time by sending a `GOODBYE` message.
> Furthermore, for `!ALLOW RESTART` and `!ALLOW CONCURRENT`, `server.done()` will check that all open connections have reached the end of the script. This means, if the client never connects to the server, this also counts as having finished the script.

Yes, we have to make sure also if the client connects to the server. Maybe I could simplify by just check if the server replied server SUCCESS once, for example.

**Comment by robsdedude** [comment]
> Yes, we have to make sure also if the client connects to the server. Maybe I could simplify by just check if the server replied server SUCCESS once, for example.

For asserting that the driver connected exactly once to a server that has `!ALLOW RESTART` or `!ALLOW CONCURRENT` in its script, I recommend something like this

```
self._routingServer1.done()
self.assertLessEqual(self._routingServer1.count_responses("<ACCEPT>"), 1)
```

Alternatively, one can consider whether the `!ALLOW` bang line(s) can be removed. Then, the second line is superfluous.

**Comment by bigmontz** [comment]
The premises of the changes were change, another PR will be open resolving the issue in another way.

**Pull request closed without merging**